src/Moq/MatcherFactory.cs

```diff
@@ -95,14 +95,30 @@ internal static class MatcherFactory
 				var convertExpression = (UnaryExpression)argument;
 				if (convertExpression.Method?.Name == "op_Implicit")
 				{
-					if (!parameter.ParameterType.IsAssignableFrom(convertExpression.Operand.Type) && convertExpression.Operand.IsMatch(out _))
+					if (convertExpression.IsMatch(out var match))
 					{
-						throw new ArgumentException(
-							string.Format(
-								Resources.ArgumentMatcherWillNeverMatch,
-								convertExpression.Operand.ToStringFixed(),
-								convertExpression.Operand.Type.GetFormattedName(),
-								parameter.ParameterType.GetFormattedName()));
+						Type foundType;
+
+						if (match.GetType().IsGenericType)
+						{
+							// If match type is `Match<int>`, foundType set to `int`
+							// Fix for https://github.com/moq/moq4/issues/1199
+							foundType = match.GetType().GenericTypeArguments[0];
+						}
+						else
+						{
+							foundType = convertExpression.Operand.Type;
+						}
+
+						if (!parameter.ParameterType.IsAssignableFrom(foundType))
```

Comment:
After inserting the missing `.Operand` (in line 98, see comment above), all of your added tests will start failing. It seems that we have the assignment compatibility check the wrong way around:

```suggestion
						if (!matchedValuesType.IsAssignableFrom(parameter.ParameterType))
```

Reply:
It is really strange that the original code fails on your machine. I think the assignment compatibility check was correct originally. I added some tests (that pass) here: https://github.com/moq/moq4/pull/1202/commits/f315474d866544f232272b37c9b0fc8e4ef650f8

when I reverse it to `!foundType.IsAssignableFrom(parameter.ParameterType)`, the tests just added fail. Thoughts?

Reply:
I am a little baffled by this actually. I'll re-run those tests tomorrow to understand better what's going on here.

Reply:
@adamfk, I took another look. By now, I am quite certain that we really need to swap `matchedValueTypes` and `parameter.ParameterType` in this test. My thinking goes like this:

We want to throw the _"unmatchable matcher"_ exception whenever we know that the matcher will receive only values that cannot possibly match its criterion. This will be the case when the type of the matcher's input values isn't assignment-compatible to whatever type the matcher is targeting (`matchedValuesType`).

What type are the matcher's input values guaranteed to have? Basically, the parameter type (`parameter.ParameterType`), because it's that parameter's argument values that'll "flow into" the matcher. The runtime enforces that all argument values are assignment-compatible to the parameter's type.

Where the types don't match, the compiler will have inserted a coercion (for primitive types) or an implicit conversion (`op_Implicit` method, for everything else), but we don't even need to know about that anymore. (That's why we may ignore the type of `convertExpression.Type`; both the compiler and the runtime guarantee that it is assignment-compatible to the parameter's type, that's the whole point of the conversion after all).

So if the type of the input values (`parameter.ParameterType`) is assignment-compatible to whatever type the matcher is targeting (`matchedValuesType`), all is well and we let everything go on. If there were an `IsAssignableTo` method, the test would thus be `parameter.ParameterType.IsAssignableTo(matchedValuesType)`, but because .NET only has `IsAssignableFrom`, the test must be swapped: `matchedValuesType.IsAssignableFrom(parameter.ParameterType)`. And because we're interested in the failure scenario, we negate using `!`.

Reply:
Sounds good. Changed in 308d711a09abe50f9374f2302e72a1997a61c404. I couldn't think of any reasonable code that would be broken by the `IsAssignableTo()` reversal.